File: src/main/java/org/c3lang/intellij/psi/impl/C3ImportPathMixinImpl.java
package org.c3lang.intellij.psi.impl;

import com.intellij.lang.ASTNode;
import org.c3lang.intellij.psi.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public abstract class C3ImportPathMixinImpl extends C3PsiElementImpl implements C3ImportPath
{
	public C3ImportPathMixinImpl(@NotNull ASTNode node)
	{
		super(node);
	}

	@Override
	public boolean endsWith(@NotNull C3Path path)
	{
		String pathText = path.getText();
		String stripped = pathText.endsWith("::") ? pathText.substring(0, pathText.length() - 2) : pathText;
		return getText().endsWith(stripped);
	}

	@Override
	public int getTextOffset()
	{
		return getFirstChild().getTextOffset();
	}

	@Override
	public @Nullable ModuleName getModuleName()
	{
		return new ModuleName(getText());
	}
}

File: src/main/java/org/c3lang/intellij/psi/impl/C3ImportPathMixinImpl.kt
package org.c3lang.intellij.psi.impl

import com.intellij.lang.ASTNode
import com.intellij.psi.util.PsiTreeUtil
import org.c3lang.intellij.dropPostfix
import org.c3lang.intellij.psi.*
import org.c3lang.intellij.psi.reference.C3ReferenceBase

abstract class C3ImportPathMixinImpl(node: ASTNode) : C3PsiElementImpl(node), C3ImportPath {

    override fun endsWith(path: C3Path): Boolean {
        return text.endsWith(path.text.dropPostfix("::"))
    }

    override fun getTextOffset(): Int {
        return firstChild.textOffset
    }

//    override fun getTextRange(): TextRange? {
//        return TextRange.create(
//            firstChild.textOffset,
//            textOffset + name.length
//        )
//    }

    override val moduleName: ModuleName?
        get() = ModuleName(text)

//    override fun getReference(): PsiReference? {
//        return PsiMultiReference(
//            arrayOf(
//                C3PathIdentExprReference(this)
//            ),
//            this
//        )
//    }

}

